matchingengine/OrderBook.go

```diff
@@ -135,7 +135,20 @@ func (ob *orderBook) AddAskOrder(ask *models.Ask) {
 	// in case of stoploss order, add it to stoploss queue and return
 	if ask.OrderType == models.StopLoss {
 		l.Debugf("Adding stopLoss with ask_id %d to the queue", ask.Id)
-		ob.askStoploss.Push(ask)
+		stockCopy, _ := models.GetStockCopy(ask.StockId)
+
+		if stockCopy.CurrentPrice <= ask.Price {
+			l.Debugf("Triggering ask %+v", ask)
+			if err := ask.TriggerStoploss(); err != nil {
+				l.Errorf("Error while changing state of %+v from StopLoss to StopLossActive", ask)
+			}
+
+			ob.asks.Push(ask)
```

Comment:
Don't directly add to `ob.asks` and `ob.depth`. Push it to the `ob.askChan`, which will take care of doing that.

Reason to do it this way is because when a limit/market order comes, we first try to fulfill as much of it as possible. Check the flow from AddAskOrder where order is limit/market type - the ask goes to `processAsk` (via `waitForOrder`). Only when the ask is fulfilled as much as possible do we put it in the queue and add to the depth.

We were treating stoploss differently. It wasn't considered active till it gets triggered during the matching. Since it wasn't considered active, it was put in its own separate queue (askStoploss). While trading, all stoplosses that should be triggered due to the new price get removed from stoploss queue and put into the ask queue.

Now, the above logic's fault is it assumes stoploss gets triggered only later and doesn't look at current price (as outlined in the issue). So all that needs to be done is to check if price should trigger the order, and if it does, treat it like a market order (pushing it to `ob.askChan`). Otherwise add to the stoploss queue